File: src/graalvm/vertx/src/main/java/com/benchmark/vertx/handlers/HealthHandler.java
package com.benchmark.vertx.handlers;

import com.benchmark.vertx.config.Config;
import com.benchmark.vertx.services.DatabaseService;
import com.benchmark.vertx.services.CacheService;
import io.vertx.core.Handler;
import io.vertx.ext.web.RoutingContext;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.time.ZoneOffset;
import java.time.ZonedDateTime;

/**
 * Health check endpoint handler.
 */
public class HealthHandler implements Handler<RoutingContext> {
    private static final Logger logger = LoggerFactory.getLogger(HealthHandler.class);

    private final Config config;
    private final DatabaseService databaseService;
    private final CacheService cacheService;

    private HealthHandler(Config config) {
        this.config = config;
        this.databaseService = new DatabaseService(config);
        this.cacheService = new CacheService(config);
    }

    public static Handler<RoutingContext> create(Config config) {
        return new HealthHandler(config);
    }

    public static Handler<RoutingContext> createHealthz() {
        return ctx -> {
            ctx.response()
                .putHeader("Content-Type", "application/json")
                .end(new io.vertx.core.json.JsonObject()
                    .put("status", "ok")
                    .encode());
        };
    }

    @Override
    public void handle(RoutingContext ctx) {
        databaseService.healthCheck()
            .compose(dbHealthy -> {
                return cacheService.ping()
                    .map(cacheHealthy -> {
                        io.vertx.core.json.JsonObject health = new io.vertx.core.json.JsonObject();
                        health.put("status", dbHealthy && cacheHealthy ? "healthy" : "unhealthy");
                        health.put("version", "1.0.0");
                        health.put("timestamp", ZonedDateTime.now(ZoneOffset.UTC).toInstant().toString());
                        health.put("database", dbHealthy ? "healthy" : "unhealthy");
                        health.put("cache", cacheHealthy ? "healthy" : "unhealthy");

                        int statusCode = (dbHealthy && cacheHealthy) ? 200 : 503;
                        ctx.response()
                            .setStatusCode(statusCode)
                            .putHeader("Content-Type", "application/json")
                            .end(health.encode());

                        return null;
                    });
            })
            .onFailure(err -> {
                logger.error("Health check failed", err);
                ctx.response()
                    .setStatusCode(503)
                    .putHeader("Content-Type", "application/json")
                    .end(new io.vertx.core.json.JsonObject()
                        .put("status", "unhealthy")
                        .put("error", err.getMessage())
                        .encode());
            });
    }
}

File: src/go/fiber/internal/handlers/health_handler.go
package handlers

import (
	"net/http"
	"time"

	"github.com/benchmark/go-fiber/internal/services"
	"github.com/gofiber/fiber/v2"
)

// HealthHandler handles health check requests
type HealthHandler struct {
	dbService  *services.DatabaseService
	cacheService *services.CacheService
}

// NewHealthHandler creates a new HealthHandler instance
func NewHealthHandler(dbService *services.DatabaseService, cacheService *services.CacheService) *HealthHandler {
	return &HealthHandler{
		dbService:     dbService,
		cacheService: cacheService,
	}
}

// HandleHealth godoc
// @Summary Health check
// @Description Check database and cache connectivity
// @Tags health
// @Accept json
// @Produce json
// @Success 200 {object} map[string]interface{} "Health status"
// @Router /health [get]
func (hh *HealthHandler) HandleHealth(c *fiber.Ctx) error {
	dbHealthy := hh.dbService.HealthCheck()
	cacheHealthy := hh.cacheService.HealthCheck()

	response := fiber.Map{
		"status":    "healthy",
		"version":   "1.0.0",
		"database":  "disconnected",
		"cache":     "disconnected",
		"timestamp": time.Now().Format(time.RFC3339),
	}

	if dbHealthy {
		response["database"] = "connected"
	}
	if cacheHealthy {
		response["cache"] = "connected"
	}

	if !dbHealthy || !cacheHealthy {
		response["status"] = "unhealthy"
		return c.Status(http.StatusServiceUnavailable).JSON(response)
	}

	return c.JSON(response)
}
